### DISCUSSION TITLE
Tangents & mBitangents

### DISCUSSION THREAD

**Discussion by RobertMtx** [assimp/assimp#4857] [Q&A] [answered]
I'm having some trouble with the `mBitangents` field of a mesh. On the longer polygons, this field does not seem to be accurate. This screen shot is of the world normals, which are calculated using tangent space:

![tangents](https://user-images.githubusercontent.com/43590381/210292579-11d4425a-bff1-4d6a-9116-e8218f10c931.png)

All of the other polygons look fine except these on the back of the couch. If I recalculate the bitangent by crossing the normal and tangent, the issue goes away entirely. While doing that fixes the problem, I'm concerned that I'm just sweeping a problem under the rug that will later show up as something nastier.

Here is the mesh file I'm loading in for diagnostics purposes:
[tangent_couch.zip](https://github.com/assimp/assimp/files/10334055/tangent_couch.zip)

Edit: sorry, here is another version of the couch with less junk in the scene:
[tangent_couch2.zip](https://github.com/assimp/assimp/files/10334083/tangent_couch2.zip)

If anyone has any idea why this is happening, or how I can fix it, please let me know. I'm assuming there is not much I could be doing wrong here because everything else is working, but please let me know if you can think of anything I should check. Thanks!

Edit: One strange thing I noticed was that this polygon has the hidden edge being flipped at some point between Blender and export. I think the retriangulation may be flipping the hidden edge, but why? Even stranger, if I run the triangulation modifier in Blender, it will also flip this hidden edge. Any ideas why everything wants to flip this polygon's edge? Or why Blender doesn't flip it before triangulation? It seems to be the only polygon that experiences this strangeness. And I've worked with hundreds of models in the past without seeing this. I don't know if this is related to the issue, but it is a strange coincidence if not.

Edit2: Well, that hidden edge being flipped had nothing to do with the issue. I tried disabling triangulation on export, which resulted in the edge remaining as it was, but the problem still remained. I also rendered another image for debugging. This render is of pure bitangents. In this image, I'm rendering the bitangent vectors exactly as they come from assimp, but with `(value+1)*0.5` to show all ranges:

![tangent_couch_just_tangents](https://user-images.githubusercontent.com/43590381/210353518-44097f89-b70d-44a3-98c8-44c43b67908b.png)

Edit3: I completely remapped the model UVs, thinking that something strange may have happened to that polygon, and for whatever reason, the problem went away. Still not really sure what happened there. I think I'm going to add a debugging test to my game engine that checks for bitangents not being equal (enough) to the cross of the normal and tangent, to help detect this type of issue in the future. If anyone has any idea what happened in my UVs that caused this, I would love to know! Thanks!

**Comment by RobertMtx** [comment]
Even with the updated UVs, I am getting some strange results from assimp's bitangent. Am I correct in assuming it should be very close to the cross product of the normal and tangent? Here is a printout of the data from one of the subsets. Many of the bitangents are very close to the cross product, but some are completely parallel to it. Can anyone explain why this happens? Should bitangents not be orthogonal to the normal and tangent?
```

DOT		ASSIMP BITANGENT	NORMAL X TANGENT (NORMALIZED)
1.000000 	0.00,-0.00,1.00		0.00,0.00,1.00
1.000000 	0.00,-0.00,1.00		0.00,0.00,1.00
1.000000 	0.00,-0.00,1.00		0.00,0.00,1.00
1.000000 	1.00,-0.00,0.00		1.00,-0.00,0.00
1.000000 	1.00,-0.00,0.00		1.00,-0.00,0.00
1.000000 	1.00,-0.00,0.00		1.00,-0.00,0.00
1.000000 	0.00,-0.00,-1.00	0.00,0.00,-1.00
1.000000 	0.00,-0.00,-1.00	0.00,0.00,-1.00
1.000000 	0.00,-0.00,-1.00	0.00,0.00,-1.00
-0.000240 	-1.00,-0.06,0.02	-0.06,0.93,-0.36
-0.000240 	-1.00,-0.06,0.02	-0.06,0.93,-0.36
-0.000348 	-1.00,-0.00,0.00	0.00,0.93,-0.36
1.000000 	0.00,-0.00,-1.00	0.00,-0.00,-1.00
1.000000 	0.00,-0.00,-1.00	0.00,-0.00,-1.00
1.000000 	0.00,-0.00,-1.00	0.00,-0.00,-1.00
0.000346 	1.00,0.00,-0.07		-0.07,0.01,-1.00
0.000346 	1.00,0.00,-0.07		-0.07,0.01,-1.00
0.000100 	0.99,0.00,-0.11		-0.11,0.01,-0.99
0.992380 	-0.97,0.24,-0.00	-0.99,0.12,0.00
0.992380 	-0.97,0.24,-0.00	-0.99,0.12,0.00
0.943110 	-0.94,0.33,0.00		-1.00,0.00,0.00
0.999804 	-0.03,0.00,-1.00	-0.01,0.00,-1.00
0.999804 	-0.03,0.00,-1.00	-0.01,0.00,-1.00
1.000000 	-0.00,-0.00,-1.00	0.00,0.00,-1.00
0.141583 	0.97,-0.00,-0.25	-0.11,0.00,-0.99
0.094343 	1.00,-0.00,-0.09	0.00,0.00,-1.00
0.094343 	1.00,-0.00,-0.09	0.00,0.00,-1.00
0.998239 	0.08,0.00,-1.00		0.03,0.00,-1.00
0.998239 	0.08,0.00,-1.00		0.03,0.00,-1.00
0.998239 	0.08,0.00,-1.00		0.03,0.00,-1.00
0.995542 	-0.99,-0.14,0.00	-0.97,-0.23,0.00
0.995542 	-0.99,-0.14,0.00	-0.97,-0.23,0.00
0.952373 	-1.00,-0.04,0.00	-0.94,-0.34,0.00
0.300969 	-0.10,1.00,0.00		-0.98,0.21,0.00
0.300969 	-0.10,1.00,0.00		-0.98,0.21,0.00
0.300969 	-0.10,1.00,0.00		-0.98,0.21,0.00
-0.005092 	-0.00,-0.00,-1.00	0.35,0.94,0.00
-0.005092 	-0.00,-0.00,-1.00	0.35,0.94,0.00
-0.005092 	-0.00,-0.00,-1.00	0.35,0.94,0.00
-0.116661 	-0.34,0.93,-0.12	-0.00,-0.00,1.00
-0.116661 	-0.34,0.93,-0.12	-0.00,-0.00,1.00
-0.116661 	-0.34,0.93,-0.12	-0.00,-0.00,1.00
1.000000 	0.00,-0.00,1.00		0.00,0.00,1.00
1.000000 	1.00,-0.00,0.00		1.00,-0.00,0.00
1.000000 	0.00,-0.00,-1.00	0.00,0.00,-1.00
-0.000023 	-0.98,-0.16,0.06	-0.18,0.92,-0.36
1.000000 	0.00,-0.00,-1.00	0.00,-0.00,-1.00
0.000840 	1.00,0.00,-0.00		0.00,0.01,-1.00
0.952373 	-1.00,0.04,-0.00	-0.94,0.34,-0.00
0.998239 	-0.08,0.00,-1.00	-0.03,0.00,-1.00
0.084585 	-1.00,-0.00,-0.08	0.00,0.00,-1.00
0.084585 	-1.00,-0.00,-0.08	0.00,0.00,-1.00
0.084585 	-1.00,-0.00,-0.08	0.00,0.00,-1.00
0.237672 	0.84,-0.00,-0.54	-0.33,0.00,-0.95
0.237672 	0.84,-0.00,-0.54	-0.33,0.00,-0.95
0.943109 	-0.94,-0.33,0.00	-1.00,0.00,0.00
0.250261 	-0.08,-1.00,0.00	-0.99,-0.17,0.00
0.250261 	-0.08,-1.00,0.00	-0.99,-0.17,0.00
0.250261 	-0.08,-1.00,0.00	-0.99,-0.17,0.00
0.012951 	0.34,0.94,-0.01		-0.00,0.00,-1.00
0.012951 	0.34,0.94,-0.01		-0.00,0.00,-1.00
0.012951 	0.34,0.94,-0.01		-0.00,0.00,-1.00
-0.004835 	0.23,-0.61,-0.76	0.26,-0.71,0.66
-0.004835 	0.23,-0.61,-0.76	0.26,-0.71,0.66
-0.004835 	0.23,-0.61,-0.76	0.26,-0.71,0.66
```

**Comment by RobertMtx** [comment, marked as answer]
Well, I learned something after some research. Tangent vectors do not necessarily need to be orthogonal or perpendicular to each other. However, if they are not, it usually means your mesh UVs are not aligned very well with the texels of the UV space. But this sort of thing happens sometimes, such as on the corners of complex shapes, where you didn't want to add an extra UV seam. The U of one vertex may be perfectly horizontal, while its V is not perfectly vertical.

So I thought I would share that, in case anyone runs into this post for the same reasons. Most likely, you need to figure out which vertices are non-orthogonal and inspect that area of the mesh. In most scenarios, it probably just needs relaxed or another seam added.

**Comment by kimkulling** [comment]
Thanks a lot for sharing the information.

**Comment by johannesugb** [comment]
In general, tangents and bitangents are not orthogonal. I have written a rather lengthy explanation in Q&A #5512 .
Chances are that you do not get the unwanted tangents/bitangents (shown in your screenshot) if you use a commit **before** https://github.com/assimp/assimp/commit/ad766cb738fbc899efe64593ff9ec2fd9c0a70a1 (which was merged in February 2022).

I see the same kind of artefacts in the Sponza model after that commit, see here:

![image](https://github.com/assimp/assimp/assets/536330/5046c8d1-bf5e-4a3f-b5cf-0c1885645a83)

The problem on the floor was not present before https://github.com/assimp/assimp/commit/ad766cb738fbc899efe64593ff9ec2fd9c0a70a1: 

![image](https://github.com/assimp/assimp/assets/536330/693e857e-c619-4cb4-9f2a-a4924cc645ae)
